docs/regalloc-demo/demo.md

```diff
@@ -0,0 +1,363 @@
+a CMake cache file that can be passed to CMake during the LLVM build
+configuration. Running the script looks like this:
+
+```bash
+mkdir tflite
+cd tflite
+/ml-compiler-opt/buildbot/build_tflite.sh
+```
+
+This script should only take a couple minutes to execute as all the libraries
+that it pulls and builds are relatively small.
+
+Now, create a new folder to do an LLVM build and configure it using CMake:
+```bash
+mkdir llvm-build
+cd llvm-build
+cmake -G Ninja \
+  -DCMAKE_BUILD_TYPE=Release \
+  -DLLVM_ENABLE_PROJECTS="clang" \
+  -C /tflite/tflite.cmake \
+  /llvm-project/llvm
+```
+
+Now you can run the actual build with the following command:
+```bash
+cmake --build .
+```
+
+## ML training
+
+All of the following example commands assume you are working from within
+your checkout of the ml-compiler-opt repository:
+
+```bash
+cd /ml-compiler-opt
+```
+
+To start off training, we need to extract a per-object bitcode/compilation
+flag corpus. This can be done with the following script:
+
+```bash
+PYTHONPATH=$PYTHONPATH:. python3 compiler_opt/tools/extract_ir.py \
+  --cmd_filter="^-O2|-O3" \
+  --input=/chromium/src/out/Release/compile_commands.json \
+  --input_type=json \
+  --llvm_objcopy_path=/llvm-build/bin/llvm-objcopy \
+  --output_dir=corpus
+```
+
+### Collect the Default Trace and Generate Vocab
+
+Before we run reinforcement training, it is best to train the model using
+behavioral cloning on the default heuristic. First off, we need to collect
+a trace of the decisions the default heuristic is making:
+
+```bash
+PYTHONPATH=$PYTHONPATH:. python3 compiler_opt/tools/generate_default_trace.py \
+  --data_path=/corpus \
+  --output_path=/default_trace \
+  --gin_files=compiler_opt/rl/regalloc/gin_configs/common.gin \
+  --gin_bindings=clang_path="'/llvm-build/bin/clang'" \
+  --sampling_rate=0.2
+```
+
+This will compile 20% of the corpus and save all of the regalloc eviction
+problem instances it encounters into the `/default_trace` file.
+
+After we have collected a default trace, an optional step is to regenerate the
+vocab that is used to normalize some of the values that get fed to the ML
+model:
+
+```bash
+rm -rf ./compiler_opt/rl/regalloc/vocab
+PYTHONPATH=$PYTHONPATH:. python3 compiler_opt/tools/sparse_bucket_generator.py \
+  --input=/default_trace \
+  --output_dir=./compiler_opt/rl/regalloc/vocab \
+  --gin_files=compiler_opt/rl/regalloc/gin_configs/common.gin
+```
+
+This isn't completely necessary as there are already default values stored
+within the repository in the `./compiler_opt/rl/regalloc/vocab` folder,
+but it definitely doesn't hurt to regenerate them. If adding/modifying
+features, it is necessary to regenerate the vocab.
+
+Now that the vocab is present (or has been regenerated) and we have a default
+trace, we can start to train the model using behavioral cloning to mimic the
+default heuristic:
+
+```bash
+PYTHONPATH=$PYTHONPATH:. python3 compiler_opt/rl/train_bc.py \
+  --root_dir=/warmstart \
```

Comment:
same here for the root_dir here, probably define a working dir as variables so the user can define it whatever they like, then have hardcoded sub-folder under the root working dir, e.g., $WORKING_DIR/warmstart, $WORKING_DIR/trace, etc

Reply:
I just modified the demo to use `$WORKING_DIR` environment variable and then based everything off of that. Probably makes things cleaner than having separate variables for everything. Thanks for the suggestion.